optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -60,21 +64,31 @@ def gaudi_mpt_attention_forward(
     if past_key_value is not None:
         if len(past_key_value) != 0:
             if token_idx is not None:
+                # copy kv to kv_cache
                 past_key_value[0].index_copy_(2, token_idx - 1, key_states)
                 past_key_value[1].index_copy_(2, token_idx - 1, value_states)
-                key_states = past_key_value[0]
-                value_states = past_key_value[1]
+
+                if cache_idx is not None:
+                    key_states = past_key_value[0][:, :, :cache_idx, :]
+                    value_states = past_key_value[1][:, :, :cache_idx, :]
+                else:
+                    key_states = past_key_value[0]
+                    value_states = past_key_value[1]
             else:
                 key_states = torch.cat([past_key_value[0], key_states], dim=2)
                 value_states = torch.cat([past_key_value[1], value_states], dim=2)
                 past_key_value = [key_states, value_states]
     else:
-        past_key_value = [
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-        ]
-        past_key_value[0][:] = key_states[:]
-        past_key_value[1][:] = value_states[:]
+        if cache_idx is not None:
```

Comment:
why is cache_idx used in this if condition? i think cache_idx is not used for the first token gen

Reply:
Only cache_idx will be passed into attention, which can be a flag of if bucket_internal was enabled. If it is enabled, computation can be reduced by assigning part of kv_cahce to key_states and value_states.